pkg/retriever/reportretriever.go

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2020 Red Hat, Inc.
+// Copyright Contributors to the Open Cluster Management project
+package retriever
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/open-cluster-management/insights-client/pkg/config"
+	"github.com/open-cluster-management/insights-client/pkg/types"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type Retriever struct {
+	CCXUrl                  string
+	client                  *http.Client
+	token                   string // token to connect to CRC
+	tokenValidationInterval time.Duration
+}
+
+type serializedAuthMap struct {
+	Auths map[string]serializedAuth `json:"auths"`
+}
+type serializedAuth struct {
+	Auth string `json:"auth"`
+}
+
+func NewRetriever() *Retriever {
+	r := &Retriever{
+		tokenValidationInterval: 5 * time.Second,
+	}
+	r.client = &http.Client{}
+	if err := r.setUpRetriever(); err != nil {
+		glog.Warningf("Unable to update mnaged clusters: %v", err)
+	}
+	return r
+}
+
+func (r *Retriever) setUpRetriever() error {
+	r.CCXUrl = config.Cfg.CCXServer
+	if !config.Cfg.UseMock {
+		go func() {
+			ticker := time.NewTicker(r.tokenValidationInterval)
+			for ; true; <-ticker.C {
+				r.StartTokenRefresh()
+			}
+		}()
+	}
+	return nil
+}
+
+func (c *Retriever) StartTokenRefresh() error {
```

Comment:
It's more efficient to "watch" the secret and update our token when changes are detected.